### PR TITLE
Ag/improvement/app refactoring

### PR DESCRIPTION
**Pull request by andriy1507**
Refactored ViewModels, Fragments and Activities. Implemented dependencies injection.

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -19,14 +19,14 @@
         android:supportsRtl="true"
         android:screenOrientation="portrait"
         android:theme="@style/MainTheme">
-        <activity android:name=".activities.StartActivity">
+        <activity android:name=".ui.activities.SplashActivity">
```
Comment (EugeneKryvoshey) [inline app/src/main/AndroidManifest.xml]: Please rename 'ui' package to 'view'.

File: app/src/main/java/com/goryachok/forecastapp/api/OpenWeatherMapAPI.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import kotlinx.coroutines.Deferred
+import retrofit2.http.GET
+import retrofit2.http.Path
+import retrofit2.http.Query
+
+
+interface OpenWeatherMapAPI {
+
+    companion object {
+        const val API_KEY = "f19f55d718abb04df88d9190337bd5b9"
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/api/OpenWeatherMapAPI.kt]: Please extract to build config.

File: app/src/main/java/com/goryachok/forecastapp/api/OpenWeatherMapAPI.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        const val DOMAIN = "https://api.openweathermap.org/"
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/api/OpenWeatherMapAPI.kt]: Please extract to build config.

File: app/src/main/java/com/goryachok/forecastapp/api/OpenWeatherMapAPI.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+//    @GET("data/2.5/weather")
+//    suspend fun getWeatherByCityAsync(
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/api/OpenWeatherMapAPI.kt]: Please either add "TODO" if you're planning to use commented code, or remove unused commented code.

File: app/src/main/java/com/goryachok/forecastapp/base/RepoViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package com.goryachok.forecastapp.base
+
+import com.goryachok.forecastapp.repository.Repository
+
+interface RepoViewModel {
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/base/RepoViewModel.kt]: Please, remove redundant abstraction.

File: app/src/main/java/com/goryachok/forecastapp/data/LocalDataSource.kt [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+class LocalDataSource @Inject constructor(context: Context) {
+
+    companion object{
+        private const val DAGGER_TAG = "DaggerDebug"
+    }
+    init {
+        Log.d(DAGGER_TAG, javaClass.name)
+    }
+
+    private val prefsName = "WEATHER_PREFERENCES"
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/data/LocalDataSource.kt]: Please move next 3 fields to constants

File: app/src/main/java/com/goryachok/forecastapp/data/LocalDataSource.kt [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    private val weatherPrefs = "LOCAL_WEATHER_DATA"
+    private val forecastPrefs = "LOCAL_FORECAST_DATA"
+
+    private val prefs = context.getSharedPreferences(prefsName, Context.MODE_PRIVATE)
+
+    fun saveData(remoteEntity: RemoteEntity) {
+        when (remoteEntity) {
+            is WeatherEntity -> {
+                val str = Gson().toJson(remoteEntity, WeatherEntity::class.java)
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/data/LocalDataSource.kt]: Try to refactor this into utils class (optional)

File: app/src/main/java/com/goryachok/forecastapp/data/LocalDataSource.kt [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            is ForecastEntity -> {
+                val str = Gson().toJson(remoteEntity, ForecastEntity::class.java)
+                prefs.edit().putString(forecastPrefs, str).apply()
+            }
+        }
+    }
+
+    fun lastFetchTime() = if (isDataAvailable()) getData(WeatherEntity::class).date else 0
+
+//    fun getWeatherData(): WeatherEntity {
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/data/LocalDataSource.kt]: Please either add "TODO" if you're planning to use commented code, or remove unused commented code.

File: app/src/main/java/com/goryachok/forecastapp/data/RemoteDataSource.kt [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+        private const val FORECAST_QUERY_PARAM = "weather"
+        private const val WEATHER_QUERY_PARAM = "forecast"
+    }
+
+    init {
+        Log.d(DAGGER_TAG, javaClass.name)
+    }
+
+//    fun getWeatherData(
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/data/RemoteDataSource.kt]: Please either add "TODO" if you're planning to use commented code, or remove unused commented code.

File: app/src/main/java/com/goryachok/forecastapp/data/RemoteDataSource.kt [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        lon: Float = 0.0F,
+        lat: Float = 0.0F
+    ): T {
+        val type = when (clazz) {
+            WeatherEntity::class.java -> WEATHER_QUERY_PARAM
+            ForecastEntity::class.java -> FORECAST_QUERY_PARAM
+            else -> throw ClassCastException()
+        }
+        lateinit var response: RemoteEntity
+        if (city != null) {
```
Comment (EugeneKryvoshey) [inline app/src/main/java/com/goryachok/forecastapp/data/RemoteDataSource.kt]: Please refactor as response assigning.

Comment (EugeneKryvoshey) [reply app/src/main/java/com/goryachok/forecastapp/data/RemoteDataSource.kt]: response = if (value != null) {
            api.getDataByCityAsync(type, city).await()
        } else {
            api.getDataByCoordinatesAsync(type, lon, lat).await()
        }

**Review by EugeneKryvoshey — CHANGES_REQUESTED** [review]

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (andriy1507) [reply app/src/main/AndroidManifest.xml]: ok

File: app/src/main/java/com/goryachok/forecastapp/api/OpenWeatherMapAPI.kt (hunk 2 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/api/OpenWeatherMapAPI.kt]: ok

File: app/src/main/java/com/goryachok/forecastapp/api/OpenWeatherMapAPI.kt (hunk 3 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/api/OpenWeatherMapAPI.kt]: ok

File: app/src/main/java/com/goryachok/forecastapp/api/OpenWeatherMapAPI.kt (hunk 4 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/api/OpenWeatherMapAPI.kt]: Removed

File: app/src/main/java/com/goryachok/forecastapp/data/LocalDataSource.kt (hunk 6 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/data/LocalDataSource.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/data/LocalDataSource.kt (hunk 8 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/data/LocalDataSource.kt]: Removed

Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/data/LocalDataSource.kt]: Removed

File: app/src/main/java/com/goryachok/forecastapp/data/RemoteDataSource.kt (hunk 10 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/data/RemoteDataSource.kt]: Done

File: app/src/main/java/com/goryachok/forecastapp/data/RemoteDataSource.kt (hunk 9 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/data/RemoteDataSource.kt]: Removed

File: app/src/main/java/com/goryachok/forecastapp/base/RepoViewModel.kt (hunk 5 above)
Comment (andriy1507) [reply app/src/main/java/com/goryachok/forecastapp/base/RepoViewModel.kt]: ok

**Review by EugeneKryvoshey — APPROVED** [review]

**Pull request merged**